File: solutions/cpp/dnd-character/2/dnd_character.cpp
#include "dnd_character.h"

#include <random>

namespace dnd_character
{
    
int roll(void) noexcept
{
    std::random_device rd{};
    std::uniform_int_distribution<int> dist(1, 6);
    return dist(rd);
}

int ability(void) noexcept
{
    int minimum = roll();
    int dice = 0;
    int sum = 0;
    for (auto i = 0; i < 3; ++i)
    {
        dice = roll();
        if (dice < minimum)
        {
            sum += minimum;
            minimum = dice;
        }
        else
        {
            sum += dice;
        }
    }
    return sum;
}
    
} // namespace dnd_character

File: solutions/java/dnd-character/2/src/main/java/DnDCharacter.java
import java.security.SecureRandom;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

record DnDCharacter(
    int getStrength, int getDexterity, int getConstitution,
    int getIntelligence, int getWisdom, int getCharisma
) {
    private static SecureRandom RNG = new SecureRandom();

    public DnDCharacter() {
        this(
            ability(), ability(), ability(),
            ability(), ability(), ability()
        );
    }

    public static int ability(List<Integer> scores) {
        return scores.stream().sorted().skip(1)
            .reduce(Integer::sum).orElseThrow();
    }

    public static int ability() {
        return ability(rollDice());
    }

    public static int roll() {
        return RNG.nextInt(1, 7);
    }

    public static List<Integer> rollDice() {
        return IntStream.generate(DnDCharacter::roll)
            .limit(4).boxed().collect(Collectors.toList());
    }

    public static int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    public int getHitpoints() {
        return 10 + modifier(getConstitution());
    }
}